javalib/src/main/scala/java/math/BigInteger.scala

```diff
@@ -0,0 +1,1500 @@
+    Conversion.toDecimalScaledString(this, 0)
+  }
+
+  /**
+   * Returns a string containing a string representation of this {@code
+   * BigInteger} with base radix. If {@code radix} is less than
+   * {@link Character#MIN_RADIX} or greater than {@link Character#MAX_RADIX}
+   * then a decimal representation is returned. The characters of the string
+   * representation are generated with method {@link Character#forDigit}.
+   *
+   * @param radix base to be used for the string representation.
+   * @return a string representation of this with radix 10.
+   */
+  def toString(radix: Int): String = {
+    Conversion.bigInteger2String(this, radix)
+  }
+
+  /**
+   * Returns a new {@code BigInteger} whose value is {@code this ^ val}.
+   * <p>
+   * <b>Implementation Note:</b> Usage of this method is not recommended as the
+   * current implementation is not efficient.
+   *
+   * @param bi value to be xor'ed with {@code this}
+   * @return {@code this ^ val}
+   * @throws NullPointerException if {@code val == null}
+   */
+  def xor(bi: BigInteger): BigInteger = Logical.xor(this, bi)
+
+  /**
+   * Returns a copy of the current instance to achieve immutability
+   */
+  def copy(): BigInteger = {
+    val copyDigits = Array.ofDim[Int](numberLength)
+    System.arraycopy(digits, 0, copyDigits, 0, numberLength)
+    new BigInteger(sign, numberLength, copyDigits)
+  }
+
+  def cutOffLeadingZeroes() {
+    while ((numberLength > 0) && (digits({numberLength  -= 1;numberLength}) == 0)) {
+    }
+    if (digits(numberLength) == 0) {
+      sign = 0
+    }
+    numberLength += 1
+  }
+
+  def equalsArrays(b: Array[Int]): Boolean = {
+    var i: Int = numberLength - 1
+    while ((i >= 0) && (digits(i) == b(i))) {
+      i -= 1
+    }
+    i < 0
+  }
+
+  def getFirstNonzeroDigit(): Int = {
+    if (firstNonzeroDigit == -2) {
+      var i: Int = 0
+      if (this.sign == 0) {
+        i = -1
+      } else {
+        i = 0
+        while (digits(i) == 0) {
+          i += 1
+        }
+      }
+      firstNonzeroDigit = i
+    }
+    firstNonzeroDigit
+  }
+
+  /**
+   * Tests if {@code this.abs()} is equals to {@code ONE}.
+   */
+  def isOne(): Boolean = {
+    (numberLength == 1) && (digits(0) == 1)
+  }
+
+  def shiftLeftOneBit(): BigInteger = {
+    if (sign == 0) this else BitLevel.shiftLeftOneBit(this)
+  }
+
+  def unCache() {
+    firstNonzeroDigit = -2
+  }
+
+  /**
+   * Puts a big-endian byte array into a little-endian applying two complement.
+   */
+  private def putBytesNegativeToIntegers(byteValues: Array[Byte]) {
+    var bytesLen = byteValues.length
+    val highBytes = bytesLen & 3
+    numberLength = (bytesLen >> 2) + (if ((highBytes == 0)) 0 else 1)
+    digits = Array.ofDim[Int](numberLength)
+    var i = 0
+    // Setting the sign
+    digits(numberLength - 1) = -1
+    // Put bytes to the int array starting from the end of the byte array
+    @inline
+    @tailrec
+    def loop: Unit = if (bytesLen > highBytes) {
```

Comment:
Define this with `()`, and call it with `()`, since it is a side-effecting function.